Apptainer container documentation (#2672)
CSCfi/csc-user-guide: docs/computing/containers/overview.md

---
description: Instructions for building and running Apptainer containers in CSC supercomputers.
---

# Apptainer containers

In this section, we provide instructions on how to build and run containers using [Apptainer](https://apptainer.org/) with fakeroot enabled in HPC clusters without unprivileged user namespaces.
We explain the special aspects of building and running containers on Puhti and Mahti clusters including how to set up the build environment, how to invoke the build commands, how to write container definition files and how to run containers.

> For general instructions about building and running containers, we recommend that users read the official [Apptainer documentation](https://apptainer.org/docs/user/main/index.html).

Reviewer: I now checked the updated intro-text, it is now perfect :+1:

Some reasons for using containers on HPC clusters are:

- Containers improve startup times and avoid I/O bottlenecks with the parallel file system, that is Lustre in Puhti and Mahti, for applications that consist of a large number of files or load many shared libraries on startup such as Python, R, and MATLAB.
  This is because Apptainer uses the Singularity Image Format (SIF) which packs the container files into a single SquashFS file.

- Running containerized software is reproducible because the container image is immutable.

- Container build definitions capture more comprehensively what is installed into the container and how it is installed.

Limitations of containers:

- Containers are not composable and they are not a substitute for package managers.
  That is, having a container with Python installed and another container with R installed, will not provide you with a container with both Python and R installed.

> ## Running containers

Reviewer 2: Is there a particular reason `apptainer_wrapper` is not recommended anymore?

Author: I can mention `apptainer_wrapper` but I think its better that users learn to use `apptainer` directly rather than the wrapper which is quite cumbersome to use.

Author: I added subsection about `apptainer_wrapper`.

Reviewer: One advantage of `apptainer_wrapper` is that it adds all the binds that you might expect to have (so that you can access user home, scratch and projappl). At least I find always having to write:

```
apptainer exec --bind="/users,/projappl,/scratch,$TMPDIR,$LOCAL_SCRATCH"` --nv ...
```
... somewhat cumbersome :-)

That said, `apptainer_wrapper` does have some things that could be done better. But we should discuss if we should improve `apptainer_wrapper` or do something else.

Using plain `apptainer` could also be made easier, for example setting default binds via `APPTAINER_BIND`. Either always, or when certain modules are loaded.

Reviewer 3: Because of the mount bindings I usually also recommend to use apptainer_wrapper..

Assume we have a container image called `container.sif`.
We can execute an arbitrary command (replace `mycommand`) inside the container using the `apptainer exec` command as follows:

```bash
apptainer exec container.sif mycommand
```

We can make directories from the host available inside the container by using bind mounts.
Specific directories that we may want to bind mount on Puhti and Mahti are the user home, projappl, scratch, and local disk spaces (see [Disk Areas](../disk.md) for more information).
Here is an example:

```bash
apptainer exec --bind="/users,/projappl,/scratch,$TMPDIR,$LOCAL_SCRATCH" container.sif mycommand
```

We can add (Nvidia) GPU support with the `--nv` flag as follows:

```bash
apptainer exec --nv container.sif mycommand
```

We can use the same flags with `apptainer run` and `apptainer shell` commands.

> ## Building container images

Reviewer 2: Link to Tykky in this section as alternative way to build Python containers?

We can build containers with Apptainer as follows:

```bash
apptainer build <options>
```

In this section, we explain how to use the command to convert existing Docker and OCI images to Singularity Image format (SIF) images, build new SIF images from definition files or develop containers interactively as modifiable (ch)root directory using a sandbox.
Also, we cover how to set the appropriate build environment and resources like memory for building on Puhti and Mahti.

### Build location

We can build containers on any node that has local disk available.
Login nodes have local disk by default.
To build on a compute node, we can reserve a job with local disk (NVMe).
Here is an example of an interactive job for building containers (change the `project_id` to your project ID):

```bash
srun \
    --account=project_id \
    --partition=interactive \
    --cpus-per-task=8 \
    --mem=16G \
    --gres=nvme:20 \
    --time=01:00:00 \
    --pty bash

> ```

Author: I can change it to sinteractive. I tend to use this notation because its standard.

---

# Apptainer containers

In this section, we provide instructions on how to build and run containers using [Apptainer](https://apptainer.org/) with fakeroot enabled in HPC clusters without unprivileged user namespaces.
We explain the special aspects of building and running containers on Puhti and Mahti clusters including how to set up the build environment, how to invoke the build commands, how to write container definition files and how to run containers.
For general instructions about building and running containers, we recommend that users read the official [Apptainer documentation](https://apptainer.org/docs/user/main/index.html).

Some reasons for using containers on HPC clusters are:

- Containers improve startup times and avoid I/O bottlenecks with the parallel file system, that is Lustre in Puhti and Mahti, for applications that consist of a large number of files or load many shared libraries on startup such as Python, R, and MATLAB.
  This is because Apptainer uses the Singularity Image Format (SIF) which packs the container files into a single SquashFS file.

- Running containerized software is reproducible because the container image is immutable.

> - Container build definitions capture more comprehensively what is installed into the container and how it is installed.

Reviewer 3: Again unsure if we want to mention it, but may-bei. A big reason in GIS to use containers is the ability to use different Linux variant and therefore Ubuntu packages.

- Containers are not composable and they are not a substitute for package managers.
  That is, having a container with Python installed and another container with R installed, will not provide you with a container with both Python and R installed.

## Running containers

### Using Apptainer directly

Assume we have a container image called `container.sif`.
We can execute an arbitrary command (replace `mycommand`) inside the container using the `apptainer exec` command as follows:

```bash
apptainer exec container.sif mycommand
```

We can make directories from the host available inside the container by using bind mounts.
On Puhti and Mahti, we can bind mount the different [Disk Areas](../disk.md) to the container as follows:

```bash
apptainer exec --bind="/users,/projappl,/scratch,$TMPDIR,$LOCAL_SCRATCH" container.sif mycommand
```

We can add Nvidia GPU support with the `--nv` flag as follows:

```bash
apptainer exec --nv container.sif mycommand
```

We can use the same flags with `apptainer run` and `apptainer shell` commands.

### Using Apptainer wrapper

Many CSC provided software environments that use containers provide access via the `apptainer_wrapper` script.

> The wrapper uses environment variables to find the path to the container image (`SING_IMAGE`) and to provide flags (`SING_FLAGS`) such as `--nv`.

Reviewer 3: Should there be a link somewhere here to all apptainer support env variables in apptainer docs?

The wrapper script automatically appends flags for common binds mounts.
We can execute commands from the container as follows:

```bash
export SING_IMAGE=/path-to/container.sif
export SING_FLAGS=""
apptainer_wrapper exec mycommand
```

> Also `apptainer_wrapper run` and `apptainer_wrapper shell` subcommand are available.

Reviewer 3: What is the difference with run and exec?

Author: I'll mention runscripts in the documentation.

Puhti and Mahti cluster set the `TMPDIR` environment variable automatically on login nodes which have local disk by default and compute nodes when local disk is reserved.
Parallel file systems such as Lustre cannot (and should not) be used as the temporary directory.

### Cache directory

Apptainer caches layers and blobs such as base images to the cache directory.
The default location is in the home directory (`$HOME/.apptainer`) which on Puhti and Mahti has a limited quota.
Thus, we may want to change the cache location to projappl to avoid filling our home directory (modify the `project_id` to your project ID).

```bash
export APPTAINER_CACHEDIR=/projappl/project_id/$USER/.apptainer
```

We can also clean the cache directory if necessary:

```bash
apptainer cache clean
```

### Virtual memory limit

Virtual memory is limited on Puhti and Mahti login nodes is quite small and should be increased to the hard limit.
Exceeding the virtual memory limit causes memory errors during build.
You can query the current virtual memory limit using `ulimit -v` and the hard limit using `ulimit -Hv`.

> On Puhti and Mahti login nodes, we can set the virtual memory limit to the hard limit as follows:

Reviewer 3: I wonder it changing the ulimit is something we recommend to do? I do not know the answer. But at least in Tykky page, if login-node memory is not enough, then the next option is to use interactive job.

Author: We can probably get rid of the need to change ulimits in the future. Ability to build on login node is a nice to have.

```bash
ulimit -v $(ulimit -Hv)
```

If your build runs out of memory, virtual memory or local disk space during the build on the login node, you should use an interactive job where virtual memory is unlimited and memory and local disk size can be configured.

### Building SIF image from existing Docker or OCI image

We can obtain existing container images from a container registry by pulling them.
Apptainer will convert them from Docker or OCI format into the Singularity Image Format (SIF).

```bash
apptainer build rockylinux.sif docker://docker.io/rockylinux/rockylinux:8
```

> You can authenticate to a private registry using `apptainer registry login` command.

Reviewer 3: Add link to Apptainer docs?

We can query information about the host system by reading the `/etc/os-release` file.

```bash
cat /etc/os-release
```

```text title="stdout"
NAME="Rocky Linux"
VERSION="8.10 (Green Obsidian)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.10"
...
```

Apptainer definition files are written using the `.def` file extension.
Here is a simple example of container definition that uses compatible base image and replaces the failing commands with the succeeding dummies:

```sh title="container.def"
Bootstrap: docker
From: docker.io/rockylinux/rockylinux:8.10

%post
    # Replace the failing commands with always succeeding dummies.
    cp /usr/bin/true /usr/sbin/useradd
    cp /usr/bin/true /usr/sbin/groupadd

    # Continue to install software into the container normally.
    dnf -y update  # would fail without the dummies
```

We can invoke Apptainer to build the container (`container.sif`) from the definition file (`container.def`) using fakeroot as follows:

```bash
apptainer build --fakeroot --bind="$TMPDIR:/tmp" container.sif container.def
```

By default Apptainer bind mounts the host's `/tmp` to `/tmp` in the build environment.

> However, the size of `/tmp` is limited on Puhti and Mahti, thus, we bind mount the local disk (`$TMPDIR`) to `/tmp` to avoid running out of memory.

Reviewer 3: Out of memory or out of temp disk space?

Author: Should be disk, thanks!

Also, we cover how to set the appropriate build environment and resources like memory for building on Puhti and Mahti.

### Choosing a Linux distribution as a base image

When selecting a base image for your container, you can choose from several Linux distributions, each with its own package manager.
For Red Hat Enterprise Linux (RHEL) based distributions that use the DNF package manager, popular options include [RedHat Universal Base Images (UBI)](https://catalog.redhat.com/en/software/base-images) available as [redhat/ubi8](https://hub.docker.com/r/redhat/ubi8) and [redhat/ubi9](https://hub.docker.com/r/redhat/ubi9), as well as community alternatives like [rockylinux](https://hub.docker.com/r/rockylinux/rockylinux) and [almalinux](https://hub.docker.com/_/almalinux).
If you prefer SUSE-based systems with the Zypper package manager, [opensuse/leap](https://hub.docker.com/r/opensuse/leap) provides a stable foundation. For Debian-based distributions using the APT package manager, both [debian](https://hub.docker.com/_/debian) and [ubuntu](https://hub.docker.com/_/ubuntu) offer well-maintained base images with extensive package repositories.

> While Apptainer allows you to build containers using any Linux distribution as the base image, building on CSC supercomputers has some limitations due from using Apptainer fakeroot mode without unprivileged user namespaces.

Reviewer: ```suggestion
While Apptainer allows you to build containers using any Linux distribution as the base image, building on CSC supercomputers has some limitations due to using Apptainer fakeroot mode without unprivileged user namespaces.
```

In this environment, certain privileged commands that are commonly executed during package installation will fail.
For example, package managers often run privileged commands, such as `useradd` and `groupadd`, as part of their installation scripts and these will fail in the fakeroot environment.

We can increase the amount of workarounds that we can use and the amount packages that install without problems by using a base image from the same family as the host system.
If you are using a base image that is not from the same family as the host system, expect that there are more packages that can't be installed successfully.
The best way to find out is to try to build the container.

You can identify your host system's Linux distribution as follows:

```bash
cat /etc/os-release
```

```text title="stdout"
NAME="Rocky Linux"
VERSION="8.10 (Green Obsidian)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.10"
...
```

> Furthermore, we should the problematic commands with dummy versions that always succeed:

Reviewer: ```suggestion
Furthermore, we can replace the problematic commands with dummy versions that always succeed:
```

### Build location

We can build containers on any node that has [local disk available](../disk.md#temporary-local-disk-areas).
Login nodes have local disk by default.
To build on a compute node, we can reserve a Slurm job with a local disk.
For example, we can reserve an interactive job with local disk (`--tmp`) as follows:

```bash
sinteractive --cores 4 --mem 4000 --tmp 10 --time 0:15:00
```

### Temporary directory

The `TMPDIR` environment variable must point to the local disk.
Apptainer will use it to identify the directory as its temporary directory when building a container.
Puhti and Mahti cluster set the `TMPDIR` environment variable automatically on login nodes which have local disk by default and compute nodes when local disk is reserved.
Lustre parallel file system cannot (and should not) be used as the temporary directory.

### Cache directory

Apptainer caches layers and blobs such as base images to the cache directory.
The default location is in the home directory (`$HOME/.apptainer`) which on Puhti and Mahti has a limited quota.

> Thus, we may want to change the cache location to projappl to avoid filling our home directory (modify the `project_id` to your project ID).

Author: Projappl is fine unless the cached images are very large.

> Thus, we may want to change the cache location to projappl to avoid filling our home directory (modify the `project_id` to your project ID).

Reviewer: Still it's kind of a temporary storage which users will forget to clean up wasting disk space ;-) Then scratch would be more logical place perhaps? But I don't have a strong opinion about this :-)

### Building SIF image from existing Docker or OCI image

We can obtain existing container images from a container registry by pulling them.
Apptainer will convert them from Docker or OCI format into the Singularity Image Format (SIF).

```bash
apptainer build rockylinux.sif docker://docker.io/rockylinux/rockylinux:8.10
```

### Building SIF image from definition file

Apptainer definition files are written using the `.def` file extension.
Here is a simple example of container definition:

```sh title="container.def"
Bootstrap: docker
From: docker.io/rockylinux/rockylinux:8.10

%post
    # Replace the failing commands with always succeeding dummies.
    cp /usr/bin/true /usr/sbin/useradd
    cp /usr/bin/true /usr/sbin/groupadd

    # Continue to install software into the container normally.
    dnf -y update  # would fail without the dummies
```

We can invoke Apptainer to build the container (`container.sif`) from the definition file (`container.def`) using fakeroot as follows:

```bash
apptainer build --fakeroot --bind="$TMPDIR:/tmp" container.sif container.def
```

> 

Reviewer: ```suggestion

See the [Examples](./examples.md) section for more examples of Apptainer definition files.
```

### Developing with interactive sandbox

We can also build Apptainer sandboxes with fakeroot.
Sandboxes are useful for interactive development of containers.
The sandbox must be created on the local disk (`$TMPDIR`), not on the Lustre parallel file system.

We can initialize a sandbox from a base image as follows:

```bash
apptainer build --fakeroot --sandbox "$TMPDIR/rockylinux" docker://docker.io/rockylinux/rockylinux:8.10
```

Then we can run a shell in the sandbox to install software into it:

```bash
apptainer shell --fakeroot --writable --contain --cleanenv --bind="$TMPDIR:/tmp" "$TMPDIR/rockylinux"
```

We can use the same tricks to replace the failing commands in the sandbox:

```bash
cp /usr/bin/true /usr/sbin/useradd
cp /usr/bin/true /usr/sbin/groupadd

> ```

Reviewer: Suggested continuation from here:

After this we can run any installation commands needed, such as `apt install` or `pip install` interactively in the sandbox shell. Once you are done, exit the shell and convert the sandbox into a `.sif` file:

```bash
apptainer build --fakeroot --bind="$TMPDIR:/tmp" container.sif "$TMPDIR/rockylinux"
```

Jaan: please double check that this is the right command, I just made it up from my head ;-)
